_posts/2022-05-14-appsync-graphql-subscriptions.md

```diff
@@ -0,0 +1,291 @@
+
+The way we had defined our subscription means that the client is interested in updates on order status when a specific orderId argument is passed through.
+
+However, if you want to enable client to subscribe to a single order `OR` all orders, you can make this argument optional by removing the exclamation point (!) from the subscription prototype.
+
+```
+type Subscription {
+    getOrderStatus($orderId: ID): Order @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+Additionally, if you want clients to explicitly subscribe to all status updates for all orders, you should remove the argument as follows:
+
+```
+type Subscription {
+    getOrderStatus: Order
+    @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+If you want to know about all orders that are created, you could do the following:
+
+```
+type Subscription {
+    getNewOrder: Order
+        @aws_subscribe(mutations: ["addOrder"])
+}
+```
+
+The golden point here to consider is that, If the filtering field is not specified in the mutation response, the subscription is not fired! If we define the addOrder mutation as bellow ( createdAt filed has intentionally been removed from mutation response):
+
+```
+export const ADD_ORDER = gql`
+  mutation addOrder(orderInput: OrderInput!) {
+    addOrder(orderInput: $orderInput) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+    }
+  }
+`;
+
+```
+
+and trying to filter the subscription on createdAt filed:
+
+```
+export const ADD_ORDER_SUBSCRIPTION = gql`
+  subscription AddOrder($createdAt: string!) {
+    getNewOrder(createdAt: $createdAt) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+
+```
+
+In this case the subscription doesn't get fired because the mutation is not requesting `createdAt` which is used for filtering in the subscription data.
+
+## Conclusion
+
+We reviewed that what are AWS AppSync, GraphQL. Followed by steps on how to implement real time session between client and the server with AWS `@aws_Subscribe` directive in AppSync. And finally wrapped it up with troubleshooting, that saves you the hours of frustration and debugging!
+
+Hopefully, by reading this article gave your enough knowledge on how subscriptions work and how simply you can implement it for your application. However AWS has a very helpful documentations for [AppSync](https://docs.aws.amazon.com/appsync/latest/devguide/what-is-appsync.html) and [Appsync Subscriptions](https://docs.aws.amazon.com/appsync/latest/devguide/aws-appsync-real-time-data.html).
+
+If you want to read more about front-end development here at Mechanical Rock I would highly recommend checking out the awesome serries of [FED Talk articles](https://mechanicalrock.github.io/2021/04/27/fed-talk-s01e01-getting-started.html) which can help you build amazing React applications hosted in the cloud.
```

Comment:
Also, sign off saying if you need help to build an awesome product, then we can always help, so get in touch 
https://www.mechanicalrock.io/our-expertise/cloud-native-solutions